### PR TITLE
Ipv4Addr and Ipv6Addr convenience constructors.

### PR DESCRIPTION
**Pull request by jcdyer**
Introduce convenience constructors for common types.

This introduces the following constructors:

* Ipv6Addr::localhost()
* Ipv6Addr::unspecified()
* Ipv4Addr::localhost()
* Ipv4Addr::unspecified()

The recently added `From` implementations were nice for avoiding the fallibility of conversions from strings like `"127.0.0.1".parse().unwrap()`, and `"::1".parse().unwrap()`, but while the Ipv4 version is roughly comparable in verbosity, the Ipv6 version lacks zero-segment elision, which makes it significantly more awkward: `[0, 0, 0, 0, 0, 0, 0, 0].into()`.  While there isn't a clear way to introduce zero elision to type that can infallibly be converted into Ipv6 addresses, this PR resolves the problem for the two most commonly used addresses, which, incidentally, are the ones that suffer the most from the lack of zero-segment elision. 

This change is dead simple, and introduces no backwards incompatibility.

See also, [this topic on the inernals board](https://internals.rust-lang.org/t/pre-rfc-convenience-ip-address-constructors/5878)

### CODE REVIEW

File: src/libstd/net/ip.rs [hunk 1]
```diff
@@ -1682,6 +1734,22 @@ mod tests {
...
+    }
+
+    #[test]
+    fn ipv6_from_contructors() {
+        assert_eq!(Ipv6Addr::localhost(), Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1));
+        assert!(Ipv6Addr::localhost().is_loopback());
+        assert_eq!(Ipv6Addr::unspecified(), Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0));
+        assert!(Ipv6Addr::unspecified().is_unspecified());
+    }
+
```
Comment (estebank) [inline src/libstd/net/ip.rs]: Trim the trailing whitespace.

File: src/libstd/net/ip.rs [hunk 2]
```diff
@@ -342,6 +342,32 @@ impl Ipv4Addr {
...
 
+    /// Creates a new IPv4 address with the address pointing to localhost: 127.0.0.1.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::net::Ipv4Addr;
+    ///
+    /// let addr = Ipv4Addr::localhost();
+    /// assert_eq!(addr, Ipv4Addr::new(127, 0, 0, 1));
```
Comment (estebank) [inline src/libstd/net/ip.rs]: missing line closing the code example:

    /// ```

File: src/libstd/net/ip.rs [hunk 3]
```diff
@@ -342,6 +342,32 @@ impl Ipv4Addr {
...
+
+    /// Creates a new IPv4 address representing an unspecified address: 0.0.0.0
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::net::Ipv4Addr;
+    ///
+    /// let addr = Ipv4Addr::unspecified();
+    /// assert_eq!(addr, Ipv4Addr::new(0, 0, 0, 0));
```
Comment (estebank) [inline src/libstd/net/ip.rs]: missing line closing the code example:

    /// ```

File: src/libstd/net/ip.rs [hunk 4]
```diff
@@ -788,6 +814,32 @@ impl Ipv6Addr {
...
 
+    /// Creates a new IPv6 address representing localhost: `::1`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::net::Ipv6Addr;
+    ///
+    /// let addr = Ipv6Addr::localhost();
+    /// assert_eq!(addr, Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1));
```
Comment (estebank) [inline src/libstd/net/ip.rs]: missing line closing the code example:

    /// ```

File: src/libstd/net/ip.rs [hunk 5]
```diff
@@ -788,6 +814,32 @@ impl Ipv6Addr {
...
+
+    /// Creates a new IPv6 address representing the unspecified address: `::`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::net::Ipv6Addr;
+    ///
+    /// let addr = Ipv6Addr::unspecified();
+    /// assert_eq!(addr, Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0));
```
Comment (estebank) [inline src/libstd/net/ip.rs]: missing line closing the code example:

    /// ```

**Review by estebank — CHANGES_REQUESTED** [review]

File: src/libstd/net/ip.rs (hunk 5 above)
Comment (jcdyer) [reply src/libstd/net/ip.rs]: All resolved. Thanks.

File: src/libstd/net/ip.rs [hunk 6]
```diff
@@ -342,6 +342,34 @@ impl Ipv4Addr {
         }
     }
 
+    /// Creates a new IPv4 address with the address pointing to localhost: 127.0.0.1.
+    ///
+    /// # Examples
+    ///
+    /// ```
```
Comment (estebank) [inline src/libstd/net/ip.rs]: I believe you need to add `# #![feature(ip)]` to this and the other new examples in order for the doc tests to work. The first `#` hides the line from the docs (which you might not want to do to make it clear that it is needed for this usage).

**Review by BurntSushi — CHANGES_REQUESTED** [review]
This looks reasonable to me, but I think these need stability attributes, and should be unstable for now.

cc @rust-lang/libs

**Comment by alexcrichton** [comment]
Looks reasonable to me as well! @jcdyer mind adding the stability attributes and filing an issue to track the stability here?

**Comment by jcdyer** [comment]
@estebank @BurntSushi I added the unstable attributes, and created a tracking issue.  I'm not entirely sure what needs to happen, but I did my best based on a quick search of other similar examples.  Let me know if anything needs to be different.

**Comment by alexcrichton** [comment]
Looks great, thanks @jcdyer!

@bors: r+

**Pull request merged**